src/test/java/org/HdrHistogram/HistogramTest.java

```diff
@@ -63,17 +64,176 @@ public void testConstructionArgumentRanges() throws Exception {
     }
 
     @Test
-    public void testLargeUnitMagnitudeDoesntInfiniteLoopInBucketCalculation() {
-        // Out of 63 bits for positive longs, subBucketCount for 5 sig  digs = 2^18 will use 19.
-        // A 45-bit lowest value will lead to a unitMagnitude of 44. 2^18 << 44 = 2^62 = (Long.MAX_VALUE / 2) + 1
-        // So, this one will not overflow in bucket calculation, and will exit normally after 1 trip through the loop.
-        assertEquals(2, new Histogram(1L << 44, Long.MAX_VALUE, 5).bucketCount);
-        // A 46-bit lowest value leads to a unitMagnitude of 45. 2^18 << 45 = 2^63 = Long.MIN_VALUE, which would cause
-        // an infinite loop if it hit the normal bucket count loop.
-        assertEquals(1, new Histogram(1L << 45, Long.MAX_VALUE, 5).bucketCount);
-        // And a 63-bit lowest value is definitely too large to fit all of subBucketCount: this can represent only
-        // one value (2^62) without overflowing.
-        assertEquals(1, new Histogram(1L << 62, Long.MAX_VALUE, 5).bucketCount);
+    public void testUnitMagnitude0IndexCalculations() {
+        Histogram h = new Histogram(1L, 1L << 32, 3);
+        assertEquals(2048, h.subBucketCount);
+        assertEquals(0, h.unitMagnitude);
+        // subBucketCount = 2^11, so 2^11 << 22 is > the max of 2^32 for 23 buckets total
+        assertEquals(23, h.bucketCount);
+
+        // first half of first bucket
+        assertEquals(0, h.getBucketIndex(3));
+        assertEquals(3, h.getSubBucketIndex(3, 0));
+
+        // second half of first bucket
+        assertEquals(0, h.getBucketIndex(1024 + 3));
+        assertEquals(1024 + 3, h.getSubBucketIndex(1024 + 3, 0));
+
+        // second bucket (top half)
+        assertEquals(1, h.getBucketIndex(2048 + 3 * 2));
+        // counting by 2s, starting at halfway through the bucket
+        assertEquals(1024 + 3, h.getSubBucketIndex(2048 + 3 * 2, 1));
+
+        // third bucket (top half)
+        assertEquals(2, h.getBucketIndex((2048 << 1) + 3 * 4));
+        // counting by 4s, starting at halfway through the bucket
+        assertEquals(1024 + 3, h.getSubBucketIndex((2048 << 1) + 3 * 4, 2));
+
+        // past last bucket -- not near Long.MAX_VALUE, so should still calculate ok.
+        assertEquals(23, h.getBucketIndex((2048L << 22) + 3 * (1 << 23)));
+        // counting by 4s, starting at halfway through the bucket
+        assertEquals(1024 + 3, h.getSubBucketIndex((2048L << 22) + 3 * (1 << 23), 23));
+    }
+
+    @Test
+    public void testUnitMagnitude4IndexCalculations() {
```

Comment:
I found it useful to stare at these various scenarios all spelled out in detail while convincing myself that the logic was correct. I can trim them down if you don't like the bulk.